### DISCUSSION TITLE
Create components and composables inside the pages folder

### DISCUSSION THREAD

**Discussion by v-romanyuk** [nuxt/nuxt#16318] [Ideas]
Hi, is there any possibility to create components/composables that are only related to some specific page inside the pages folder to keep everything related at one place.
Right now we're extending the following hooks to make it possible:
- pages:extend - here we remove all the files from _components directory that can be placed at any level of pages folder to make them not to be treated as routes.
- components:dirs - here we extend components directories that are placed in pages folder using the same _components pattern
- imports:extend: here we add ts files with some specific pattern to make them to be auto impotrable.

@danielroe it would be really nice to hear your thoughts on it. Is this possible to do in any kind of native way without any workarounds?

**Comment by wobsoriano** [comment]
You are 100% looking for this https://github.com/davestewart/nuxt-areas

But support for [Nuxt 3](https://github.com/davestewart/nuxt-areas/issues/7) is a WIP

**Reply by v-romanyuk** [reply]
Well, I actually was able to achieve this using modules in conjunction with hooks, but as I said maybe nuxt team has a better approach. Thanks for sharing nuxt-areas though.

**Comment by TheAlexLichter** [comment]
Did you check https://github.com/nuxt/framework/issues/3222 / `extends`?

**Reply by v-romanyuk** [reply]
Yeah with this approach you have to create nuxt.config.ts for every folder which is not very convenient. Furthermore it doesn't work with root 'pages' folder.

**Comment by igoohd** [comment]
In Nuxt 3, you can config a pattern to prevent it from being registered as a route like an underscore (`_`) .

nuxt.config.ts:
`export default defineNuxtConfig({
...,
ignore: ['pages/**/_*']
})`

#### Example:
pages/
├── _components/
│└── MyComponent.vue
├── _composables/
│ └── useMyComposable.ts
├── index.vue
└── _types.ts

Any file or folder prefixed with `_` in pages directory is ignored, allowing you to keep your directory structure organized within the `pages` folder.

Docs: https://nuxt.com/docs/api/nuxt-config#ignore and https://nuxt.com/docs/guide/directory-structure/nuxtignore

**Reply by Kokleng-Dev** [reply]
it work well, but hot reload not work when i change something inside _component.
i need to re-run it every time. how to solve it? thank you.

**Reply by igoohd** [reply]
Sorry for the delay—hopefully this is still helpful!

I ended up trying a different approach than I originally suggested. Here's what worked for me in nuxt.config.ts:


```
export default defineNuxtConfig({
  hooks: {
    'pages:extend'(pages) {
      /**
       * Recursively filters out pages based on filename and folder naming patterns.
       * 
       * @param pages - The array of NuxtPage objects to be filtered
       * @param filePattern - A RegExp pattern to match unwanted files (e.g., .ts files)
       * @param folderPattern - A RegExp pattern to match unwanted folders (e.g., folders starting with '_')
       * @returns A new filtered array of NuxtPage objects
       */
      function filterPages(
        pages: NuxtPage[],
        filePattern: RegExp,
        folderPattern: RegExp
      ): NuxtPage[] {
        const filteredPages: NuxtPage[] = []

        pages.forEach((page) => {
          // Check if the page has a corresponding file path
          if (page.file) {
            // Exclude pages that match either the file pattern or the folder pattern
            if (!filePattern.test(page.file) && !folderPattern.test(page.file)) {
              // Recursively filter child pages
              page.children = filterPages(page.children || [], filePattern, folderPattern)
              filteredPages.push(page)
            }
          } else {
            // If the page has no file (likely a parent page), still filter its children
            page.children = filterPages(page.children || [], filePattern, folderPattern)
            filteredPages.push(page)
          }
        })

        return filteredPages
      }

      // Pattern to ignore TypeScript files directly inside the pages directory
      const tsFilePattern = /\.ts$/

      // Pattern to ignore folders that start with an underscore (e.g., /pages/_internal)
      const underscoreFolderPattern = /\/_/

      // Apply the filter in-place on the original pages array
      pages.splice(
        0,
        pages.length,
        ...filterPages(pages, tsFilePattern, underscoreFolderPattern)
      )
    },
  },
})

```
